src/main/java/org/opensearch/agent/tools/DataDistributionTool.java

```diff
@@ -0,0 +1,1298 @@
+
+        if (allKeys.isEmpty()) {
+            return Double.NEGATIVE_INFINITY;
+        }
+        return allKeys.stream().mapToDouble(key -> {
+            double selectionVal = selectionDist.getOrDefault(key, 0.0);
+            double baselineVal = baselineDist.getOrDefault(key, 0.0);
+            return selectionVal - baselineVal;
+        }).max().orElse(Double.NEGATIVE_INFINITY);
+    }
+
+    /**
+     * Extracts field names from sample data when mapping is not available
+     *
+     * @param data Sample data to analyze
+     * @return List of field names suitable for analysis
+     */
+    private List<String> getFieldsFromData(List<Map<String, Object>> data) {
+        if (data.isEmpty()) {
+            return List.of();
+        }
+
+        Set<String> allFields = new HashSet<>();
+        for (Map<String, Object> doc : data) {
+            allFields.addAll(doc.keySet());
+        }
+
+        // Filter out timestamp and other non-useful fields
+        return allFields
+            .stream()
+            .filter(field -> !field.equals("@timestamp") && !field.equals("_id") && !field.equals("_index"))
+            .filter(field -> {
+                // Check cardinality - exclude high cardinality fields
+                Set<String> values = new HashSet<>();
+                for (Map<String, Object> doc : data) {
+                    Object value = doc.get(field);
+                    if (value != null) {
+                        values.add(String.valueOf(value));
+                    }
+                }
+                int cardinality = values.size();
+                return cardinality > 0 && cardinality <= Math.max(10, data.size() / 2);
+            })
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Gets number fields from field type mappings
+     *
+     * @param fieldTypes Map of field names to their types
+     * @return Set of number field names
+     */
+    private Set<String> getNumberFields(Map<String, String> fieldTypes) {
+        return fieldTypes
+            .entrySet()
+            .stream()
+            .filter(entry -> NUMBER_FIELD_TYPES.contains(entry.getValue()))
+            .map(Map.Entry::getKey)
+            .collect(Collectors.toSet());
+    }
+
+    /**
+     * Groups numeric keys and merges counts
+     *
+     * @param selectionDist Selection distribution
+     * @param baselineDist Baseline distribution
+     * @return Grouped distributions
+     */
+    private GroupedDistributions groupNumericKeys(Map<String, Double> selectionDist, Map<String, Double> baselineDist) {
+        Set<String> allKeys = new HashSet<>(selectionDist.keySet());
+        allKeys.addAll(baselineDist.keySet());
+
+        if (allKeys.size() <= 10 || !allKeys.stream().allMatch(key -> {
+            try {
+                Double.parseDouble(key);
+                return true;
+            } catch (NumberFormatException e) {
+                return false;
+            }
```

Comment:
I think apache common lang `NumberUtils` can used to check if a give string a number, take a look on that.

Reply:
It seems `!allMatch` equals to `anyMatch`, and anyMatch might have better performance as it can stop early once found any element not match.